### PR TITLE
Implement middleware for default HEAD implementation

### PR DESCRIPTION
**Pull request by rossabaker**

### CODE REVIEW

**Comment by rossabaker** [comment]
Now that I've done it, with more MiMa exceptions, this could technically be put on release-0.14.x.

**Comment by rossabaker** [comment]
Also seems like this could be worked into the example service somewhere.  I'd like to come up with a recommended set of middleware for things that make a good HTTP citizen, and this qualifies.

File: server/src/main/scala/org/http4s/server/middleware/DefaultHead.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+package middleware
+
+import scalaz.concurrent.Task
+import org.http4s.server.syntax._
+
+/** Handles HEAD requests as a GET without a body.
+  * 
+  * If the service returns the fallthrough response, the request is resubmitted
+  * as a GET.  The resulting response's body is killed, but all headers are
+  * preserved.  This is a naive, but correct, implementation of GET.  Routes
```
Comment (bryce-anderson) [inline server/src/main/scala/org/http4s/server/middleware/DefaultHead.scala]: Did you mean 'This is a naive, but correct, implementation of _HEAD_'?

File: server/src/test/scala/org/http4s/server/middleware/DefaultHeadSpec.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      Ok().putHeaders(Header("X-Handled-By", "HEAD"))
+  })
+
+  "DefaultHead" should {
+    "honor HEAD routes" in {
+      val req = Request(Method.HEAD, uri = uri("/special"))
+      service(req).map(_.headers.get("X-Handled-By".ci).map(_.value)) must returnValue(Some("HEAD"))
+    }
+
+    "return truncated body of corresponding GET on fallthrough" in {
```
Comment (bryce-anderson) [inline server/src/test/scala/org/http4s/server/middleware/DefaultHeadSpec.scala]: Could we check the `Content-Length` header as well? For that matter, we could check that the `Response` objects are the exact same except the `body`.

**Pull request merged**